Helper/Data.php

```diff
@@ -130,6 +176,213 @@ public function getSellerIdByProductId($productId = '')
+    {
+        try {
+            $isFinanced = $params['isFinanced'] ? 1 : 0;
+            $selectedPaymentMethod = (float)$params['selectedPaymentMethod'];
+            $orderPayment = $order->getPayment();
+            $orderPayment
+                ->setAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_FINANCED, $isFinanced)
+                ->setAdditionalInformation(BalancepayMethod::
+                BALANCEPAY_SELECTED_PAYMENT_METHOD, $selectedPaymentMethod);
+            $orderPayment->save();
+            $order->save();
+            return true;
+        } catch (\Exception $e) {
+            $this->balancepayConfig->log($e->getMessage());
+        }
+    }
+
+    /**
+     * @param $params
+     * @param $order
+     * @return void
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function processChargedWebhook($params, $order)
+    {
+        try {
+            $chargeId = (string)$params['chargeId'];
+            $amount = (float)$params['amount'];
+            $orderPayment = $order->getPayment();
+
+            if (\strpos($orderPayment
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID), $chargeId) === false) {
+                if (!$orderPayment
+                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)
+                    && round((float)$order->getBaseGrandTotal()) !== round($amount)) {
+                    $orderPayment->setIsFraudDetected(true)->save();
+                    $order->setStatus(Order::STATUS_FRAUD)->save();
+                    throw new LocalizedException(new Phrase("The charged amount doesn't match the order total!"));
+                }
+
+                $orderPayment
+                    ->setTransactionId($orderPayment
+                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHECKOUT_TRANSACTION_ID))
+                    ->setIsTransactionPending(false)
+                    ->setIsTransactionClosed(true)
+                    ->setAdditionalInformation(
+                        BalancepayMethod::BALANCEPAY_CHARGE_ID,
+                        $orderPayment->getAdditionalInformation(
+                            BalancepayMethod::BALANCEPAY_CHARGE_ID,
+                            $chargeId
+                        ) . " \n" . $chargeId
+                    );
+
+                if (!$orderPayment
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)) {
+                    $orderPayment->capture(null);
+                }
+                $orderPayment->save();
+                $order->save();
+                return true;
+            } elseif ($chargeId !== (string)$order->getPayment()
+                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID)) {
+                throw new LocalizedException(new Phrase("Charge ID mismatch!"));
+            }
```

Comment:
move this to be a negative condition before its related `if` statement to keep code cleaner:

```
$orderChargeId = (string)$order->getPayment()
                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID);
if ($chargeid !== $orderChargeId) {
    throw new LocalizedException(new Phrase("Charge ID mismatch!"));
}

if (\strpos($orderPayment
                    ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_CHARGE_ID), $chargeId) === false) {
                if (!$orderPayment
                        ->getAdditionalInformation(BalancepayMethod::BALANCEPAY_IS_AUTH_CHECKOUT)
                    && round((float)$order->getBaseGrandTotal()) !== round($amount)) {
    // ...
}
```

Reply:
This we cannot move up as charge id is not yet set to the payment object.